File: src/api/settings/dto/updateSettingsRequestDto.ts
import { AiProvider, AiProviderSettings, Font, Theme } from "./settingsDto";

export default interface UpdateSettingsRequestDto {
	baseDatabaseDirectory: string | null;

	theme: Theme | null;
	font: Font | null;
	fontHeadings: Font | null;
	fontMonospace: Font | null;
	zoomPercentage: number | null;
	autoSync: boolean | null;
	trashRetentionDays: number | null;

	enableAi: boolean | null;
	aiProvider: AiProvider | null;
	ollama: AiProviderSettings | null;
	openai: AiProviderSettings | null;
}

/**
 * Builds an {@link UpdateSettingsRequestDto} that leaves every field unchanged
 * (all `null`), overriding only the fields provided. Use this so callers only
 * specify the settings they actually want to change.
 */
export function buildUpdateSettingsRequest(
	overrides: Partial<UpdateSettingsRequestDto>,
): UpdateSettingsRequestDto {
	return {
		baseDatabaseDirectory: null,
		theme: null,
		font: null,
		fontHeadings: null,
		fontMonospace: null,
		zoomPercentage: null,
		autoSync: null,
		trashRetentionDays: null,
		enableAi: null,
		aiProvider: null,
		ollama: null,
		openai: null,
		...overrides,
	};
}

File: src-tauri/src/settings/dto/update_settings_request_dto.rs
use std::path::PathBuf;

use serde::{Deserialize, Serialize};

use crate::settings::value_objects::{
    ai_provider::AiProvider, ai_provider_settings::AiProviderSettings, font::Font,
    settings_profile::SettingsProfile, theme::Theme,
};

#[derive(Default, Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateSettingsRequestDto {
    pub base_database_directory: Option<PathBuf>,
    pub profile: Option<SettingsProfile>,

    pub theme: Option<Theme>,
    pub font: Option<Font>,
    pub font_headings: Option<Font>,
    pub font_monospace: Option<Font>,
    pub zoom_percentage: Option<f64>,
    pub auto_sync: Option<bool>,
    pub trash_retention_days: Option<u32>,

    pub enable_ai: Option<bool>,
    pub ai_provider: Option<AiProvider>,
    pub ollama: Option<AiProviderSettings>,
    pub openai: Option<AiProviderSettings>,
}
